### DISCUSSION TITLE
magit status buffer show unpushed commits for non-checked-out branches?

### DISCUSSION THREAD

**Discussion by ville-h** [magit/magit#5015] [Ideas]
Currently magit status buffer appears to show unpushed commits only for the checked out branch? How do I get it to show this for all the branches which would have that section showing would the branch be the checked-out branch?

I may work on multiple branches simultaneously and for whatever reason may have to delay pushing any one of them. It would be nice to have a visible reminder showing that there are specific branches that have unpushed commits.

**Comment by tarsius** [comment]
IMO that's information that does not belong in the status buffer. What you propose might work well if you have two or three such "other branches", but not so well if there's a hundred.

You can get information like this in the "Refs" buffer, which you can get to with `y`.

```
Branches (3)
  idea20>      <23main [maybe] magit-insert-section: Purge last generation markers
@ main1>          origin/main magit-read-char-case: Always separate one choice with a "or"
  sad1>         <3main Revert "Use seq-keep instead of delq and mapcar"
+ 73bf95d468 Revert "Use seq-keep instead of delq and mapcar"
```

Here `sad` is 1 commit ahead of its upstream `main` and 3 behind. The respective section is expanded showing the commit that it is ahead (using `git cherry -v`).

**Comment by ville-h** [comment]
Yes it could be unusable with 100s, 1000s, ... branches that have things to push. However that is not the usual situation for me. It's a handful of branches at any given time. Right now magit does not seem to offer the choice for the user to get this information in a convenient way. Allow the users to decide if it's information they wish to see on the status buffer or not by providing an implementation of a function they may add to the `magit-status-section-hooks`.

**Reply by tarsius** [reply]
Here's some code to do that:

```elisp
(defun magit-insert-all-the-branches ()
  (dolist (branch (delete (magit-get-current-branch)
                          (magit-list-local-branch-names)))
    (setq branch (propertize branch 'font-lock-face 'magit-branch-local))
    (when-let ((upstream (magit-get-upstream-branch branch)))
      (when (magit-git-success "rev-parse" (format "%s@{upstream}" branch))
        ;; Based on `magit-insert-unpushed-to-upstream'.
        (magit-insert-section
          (unpushed (eval (format "%s@{upstream}..%s" branch branch)) t)
          (magit-insert-heading
            (format (propertize "%s unmerged into upstream %s."
                                'font-lock-face 'magit-section-heading)
                    branch upstream))
          (magit--insert-log nil
            (format "%s@{upstream}..%s" branch branch)
            magit-buffer-log-args)
          (magit-log-insert-child-count))
        ;; Based on `magit-insert-unpulled-from-upstream'.
        (magit-insert-section
          (unpulled (eval (format "%s..%s@{upstream}" branch branch)) t)
          (magit-insert-heading
            (format (propertize "%s unpulled from upstream %s."
                                'font-lock-face 'magit-section-heading)
                    branch upstream))
          (magit--insert-log nil
            (format "%s..%s@{upstream}" branch branch)
            magit-buffer-log-args)
          (magit-log-insert-child-count))))))

(magit-add-section-hook 'magit-status-sections-hook
                        #'magit-insert-all-the-branches nil t)
```

But I don't think this should be added to Magit.

**Comment by ville-h** [comment]
There's a chance we're discussing two different things or the above function does not work. So here are some pictures that hopefully clarify what I am talking about.

![emacs_magit_status_buffer_000](https://github.com/magit/magit/assets/8824715/a19c465a-bcb3-488d-a095-cba5dba32153)

Above is a magit status buffer when `main` is checked out. `<-1` indicating the "unpushed" information for the current branch.

![emacs_magit_status_buffer_001](https://github.com/magit/magit/assets/8824715/2ff95aa4-40ab-427a-a1ed-3a4fcb05829d)

Above is a magit satus buffer when `type_check` is checked out. `<-3` indicating that no "unpushed" information is shown. This happens without or with the above code snippet. I would like that the information from the first picture indicated by `<-1` would be displayed here as well when `type_check` is checked out.

 Now caveat that I am not a lisp programmer so I don't really understand what your function does, but I was able to add `(message "XXX")` in the function definition. I can see that the message is printed at least. So perhaps it's being called at the appropriate time, but doesn't quite produce the kind of output I was hoping for.

**Reply by tarsius** [reply]
The code I presented implements this for the "upstream" branch, but it seems you don't configure that and only set a branch's "push target" branch.

[But there also was a bug in the above code, so start by copying the updated version.]

If you set both, then it looks like this for a feature branch:

```
Head:     notif [wip] discussions
Merge:    main manual: Document forge-alist
Push:     origin/notif [wip] show repo
```

and for the main branch:

```
Head:     main ghub--graphql-walk-response: Cosmetics
Merge:    origin/main Place comma after "e.g."
Push:     origin/main Place comma after "e.g."
```

(These examples are from different repositories. The point is that the "upstream" and "push target" are usually the same for the main branch.)

You can learn about this at https://magit.vc/manual/magit/The-Two-Remotes.html.

IMO it would be better if you made sure that the upstream is set (but continue to also set the push target), but here's an adaption for the push targets:

I don't know why the upstream isn't set for your branches, that's something that usually gets set when you create a new branch.

IMO it would be better if you made sure that the upstream is set (but continue to also set the push target), but here's an adaption for the push targets:

```elisp
(defun magit-insert-all-the-branches ()
  (dolist (branch (delete (magit-get-current-branch)
                          (magit-list-local-branch-names)))
    (setq branch (propertize branch 'font-lock-face 'magit-branch-local))
    (when-let ((target (magit-get-push-branch branch)))
      (progn
        ;; Based on `magit-insert-unpushed-to-pushremote'.
        (magit-insert-section
          (unpushed (eval (format "%s..%s" target branch)) t)
          (magit-insert-heading
            (format (propertize "%s unmerged into target %s."
                                'font-lock-face 'magit-section-heading)
                    branch target))
          (magit--insert-log nil
            (format "%s..%s" target branch)
            magit-buffer-log-args)
          (magit-log-insert-child-count))
        ;; Based on `magit-insert-unpulled-from-pushremote'.
        (magit-insert-section
          (unpulled (eval (format "%s..%s" branch target)) t)
          (magit-insert-heading
            (format (propertize "%s unpulled from target %s."
                                'font-lock-face 'magit-section-heading)
                    branch target))
          (magit--insert-log nil
            (format "%s..%s" branch target)
            magit-buffer-log-args)
          (magit-log-insert-child-count))))))

(magit-add-section-hook 'magit-status-sections-hook
                        #'magit-insert-all-the-branches nil t)
```
